### ISSUE TITLE
Submodule candidates

### ISSUE THREAD

**Issue by tresf** [LMMS/lmms#3963]
#### Candidates

| PR | Plugin | Upstream | Comments |
|-------|-------|------------|-----------|
| #3873 | core/rpmalloc | github.com/rpmalloc/rpmalloc | ✅  Done |
| #3931 | ladspa/swh | github.com/swh/ladspa | ✅ Done |
| #3960  | ladspa/tap | github.com/tomszilagyi/tap-plugins | ✅ Done |
| #3940 #3987  | ladspa/calf | github.com/lmms/veal  |  ✅ 🍴 Done.  Upstream API changed, need to use fork. |
| #4027 (WIP) | ladspa/caps | quitte.de/dsp/ | ⌛ In progress using moddevices Caps LV2 mirror  |
| #5755  | ladspa/cmt | github.com/lmms/cmt | ✅ 🍴 Done (our own fork, no upstream VCS)  |
| #3946   | plugins/papu | bitbucket.org/mpyne/game-music-emu | ✅  Done |
| #3962  | plugins/opl2 | github.com/adplug/adplug | ✅ Done  |
| #5724  | plugins/sid | github.com/simonowen/resid | ✅ Done.  |
| #3962  | plugins/zynaddsubfx | sourceforge.net/p/zynaddsubfx | ✅ 🍴 Done.  Upstream API changed.  Need to use fork.  |
| #3965 | plugins/xpressive | github.com/ArashPartow/exprtk | ✅  Done.  ~~Requires patch file for MacOS.~~ Not anymore, upstream patched.  |
| #5713 | plugins/carlabase | github.com/falktx/carla | ✅  Done (weak linking, like weakjack) |
| #3970  | audio/weakjack | github.com/x42/weakjack | ✅ Done.  |
| #5716 | 3rdparty/jackaudio/jack | github.com/jackaudio/headers | ✅ Done. |
| #4588 | wiki | github.com/lmms/lmms.wiki.git | ✅ Done. | 

#### Potential candidates

| Plugin | Upstream | Comments |
|--------|----------|------------|
| plugins/ReverbSC| github.com/csound/csound | ⚠️  42MB library, better candidate for `subtree`? |

Are there any I'm missing?

Related #3957

**Comment by JohannesLorenz** [comment]
While I welcome a clonable wiki, I'm not sure if a submodule is justified. LMMS is already full of submodules, making cloning a hell if your internet is limited. Though, some users may user the shallow clones.

It might also depend on the size of that submodule (think about images, videos etc). How large would that submodule be?

Any more opinions here?

**Comment by tresf** [comment]
> How large would that submodule be?

277 KB

**Comment by tresf** [comment]
> Think about images, videos etc

GitHub doesn't allow these in the wiki.  If you see one, it was pasted into a bug report and cross-linked.

**Comment by tresf** [comment]
This is what it would look like.  Interestingly enough, GitHub won't allow you to browse the files in their raw form. https://github.com/tresf/lmms/tree/wiki/doc

**Comment by tresf** [comment]
Thanks.  Created the PR #4588, updated the above checklist.

**Comment by trebmuh** [comment]
CMT had a release in January this year. [See there](http://www.ladspa.org/cmt/changes.html).

**Comment by zonkmachine** [comment]
**plugins/SID**

Updated version of resid here: https://sourceforge.net/p/vice-emu/code/HEAD/tree/trunk/vice/src/resid/

**Comment by tresf** [comment]
> CMT had a release in January this year

It seems that Debian's Salsa project has a viable mirror: https://salsa.debian.org/multimedia-team/cmt.

> Updated version of resid here: https://sourceforge.net/p/vice-emu/code/HEAD/tree/trunk/vice/src/resid/

It appears they're still using SVN, so we can't submodule it unless we fork it.  Git has a command `git svn`, but it's not installed by default and seems to have some [nuances with submodules](https://stackoverflow.com/a/5345092/3196753).  I actually started the `git svn clone` process with reside and it's been running for a long, long time, so I'd hate to chain it into our build system.

https://github.com/simonowen/resid seems to be updated semi-regularly, we could potentially clone that, but it's a single-maintainer repo, so it's likely to bit rot over time.

There's also an [old mirror here](https://github.com/svn2github/vice-emu), but the service that was cloning it -- svn2github -- has since been discontinued, so it would require some involvement to get that resurrected.  I'd happily convert the plugin over to @simonowen's copy if that's desired.  Pinging him incase he knows a bit more about reliably git repos. :)

**Comment by simonowen** [comment]
@tresf : The changes I made to reSID were mostly to move it from automake to CMake, and clean up a few build warnings. Version 0.16 was the latest version I could find to use as a base for my repositiory, which dated back to 2004. Few other projects seemed to use reSID, and those that did usually included the code directly in their repositories, which wasn't convenient for me.

I'm not planning on making any functional changes to the API or SID emulation. The only additional changes I was considering were to add Travis/AppVeyor CI build testing, and change it to use a project-specific CMake symbol to control static builds (instead of relying on the global `BUILD_SHARED_LIBS`).

It should be simple to chain my respository into an existing project that already uses CMake. If you do find a better repository source, I may well move over to it myself  :)

**Comment by tresf** [comment]
@simonowen thanks kindly for the explanation.  We'll point at yours then.  PR opened #5724.  Checklist updated.

**Comment by tresf** [comment]
> | Plugin | Upstream | Comments |
> |--------|----------|------------|
> | plugins/ReverbSC| github.com/csound/csound | ⚠️  42MB library, better candidate for `subtree`? |

I made an attempt to bring the `csound/csound` library in as a submodule, but I'm struggling to do so.  It appears the `base.h` that we currently include is a rather generic soundpipe/io helper and `csound` uses its own.  Some `struct` signatures and names have changed as well.  When combined this is far beyond my abilities.  If someone is willing to run with this, I welcome it. :)

**Comment by tresf** [comment]
I'm inclined to host our own copy of the CMT plugins.  Perhaps this is overly ambitious, but crawling GitHub for mirrors landed me only [one modern copy](https://github.com/sjfloat/cmt).  Doing this properly, I'd prefer to create the repo and make version releases as they match upstream.  That way we'll have the change history which will help sorting any regressions we'd hit as a result of an upgrade.

This also means we need to validate our CMT against upstream and make sure there aren't any unaccounted changes.  I'll start this process now over at https://github.com/lmms/cmt .  @sjfloat once this is created, feel free to fork it if you want. :)

**Comment by tresf** [comment]
This thread has run its course.  Thanks to all involved that helped.  CAPS is the final remaining submodule, but that can be tracked individually.

Special thanks to @JohannesLorenz for backporting Calf from LV2 to LADSPA as well as the various upstream maintainers making this effort possible.  This effort makes the LMMS codebase smaller while encouraging upstream patching whenever possible.

I've attached a summary of this effort below illustrating how we've removed 279,628 lines of code and removed 610 tracked source files.

> ![Source Lines Removed By Submodules (2)](https://user-images.githubusercontent.com/6345473/98770093-6e32bd00-23af-11eb-8479-087530160216.png)

> ![Source Files Removed By Submodules (1)](https://user-images.githubusercontent.com/6345473/98770097-6ffc8080-23af-11eb-9065-628fd77fad01.png)
